Kernel/FileSystem/FUSE/Inode.cpp

```diff
@@ -0,0 +1,334 @@
+        fuse_read_in payload {};
+        payload.fh = id;
+        payload.offset = offset + nread;
+        payload.size = chunk_size;
+        auto response = TRY(fs().m_connection->send_request_and_wait_for_a_reply(FUSEOpcode::FUSE_READ, nodeid, { &payload, sizeof(payload) }));
+
+        fuse_out_header* header = bit_cast<fuse_out_header*>(response->data());
+        if (header->error)
+            return Error::from_errno(-header->error);
+
+        u32 data_size = min(target_size, header->len - sizeof(fuse_out_header));
+        if (data_size == 0)
+            break;
+
+        u8* data = bit_cast<u8*>(response->data() + sizeof(fuse_out_header));
+
+        TRY(buffer.write(data, nread, data_size));
+        nread += data_size;
+        target_size -= data_size;
+    }
+
+    TRY(try_flush(id));
+    TRY(try_release(id, false));
+
+    return nread;
+}
+
+ErrorOr<size_t> FUSEInode::write_bytes_locked(off_t offset, size_t size, UserOrKernelBuffer const& buffer, OpenFileDescription*)
+{
+    VERIFY(m_inode_lock.is_locked());
+    VERIFY(!is_directory());
+    VERIFY(offset >= 0);
+
+    constexpr size_t max_write_size = 0x21000 - sizeof(fuse_in_header) - sizeof(fuse_write_in);
+    u64 id = TRY(try_open(false, O_WRONLY));
+    u32 nodeid = identifier().index().value();
+
+    size_t nwritten = 0;
+    while (size) {
+        size_t chunk_size = min(size, max_write_size);
+        auto request_buffer = TRY(KBuffer::try_create_with_size("FUSE: Write buffer"sv, sizeof(fuse_write_in) + chunk_size));
+        fuse_write_in* write_header = bit_cast<fuse_write_in*>(request_buffer->data());
+        write_header->fh = id;
+        write_header->offset = offset + nwritten;
+        write_header->size = chunk_size;
+        TRY(buffer.read(request_buffer->data() + sizeof(fuse_write_in), nwritten, chunk_size));
+
+        auto response = TRY(fs().m_connection->send_request_and_wait_for_a_reply(FUSEOpcode::FUSE_WRITE, nodeid, request_buffer->bytes()));
+        if (response->size() < sizeof(fuse_out_header) + sizeof(fuse_write_out))
+            return Error::from_errno(EIO);
+
+        fuse_out_header* header = bit_cast<fuse_out_header*>(response->data());
+        if (header->error)
+            return Error::from_errno(-header->error);
+
+        fuse_write_out* write_response = bit_cast<fuse_write_out*>(response->data() + sizeof(fuse_out_header));
+
+        nwritten += write_response->size;
+        size -= write_response->size;
+    }
+
+    TRY(try_flush(id));
+    TRY(try_release(id, false));
+
+    return nwritten;
+}
+
+InodeMetadata FUSEInode::metadata() const
+{
+    InodeMetadata metadata;
+    metadata.inode = identifier();
+    u32 id = identifier().index().value();
+
+    fuse_getattr_in payload {};
+
+    auto response_or_error = fs().m_connection->send_request_and_wait_for_a_reply(FUSEOpcode::FUSE_GETATTR, id, { &payload, sizeof(payload) });
+
+    // FIXME: Propagate errors somehow.
+    if (response_or_error.is_error())
```

Comment:
We could technically just return invalid metadata (all zeroed-out). I think that we already do so for filesystems like `Plan9`?